docs/specs/webhook_update.md

```diff
@@ -0,0 +1,87 @@
+## Update webhooks
+
+### Concepts
+
+Instead of having before or after hooks, introduce new concept: blocking and non-blocking hooks.
+
+Actually blocking hooks are the before books and non-blocking is after. But they don't need to be exists in pair anymore.
+
+Blocking hooks accept below response body, operation will be aborted when `is_allowed` is `false`. `title` and `reason` will be shown in auth ui.
+
+```json5
+{
+  "is_allowed": false,
+  "title": "any title",
+  "reason": "any string",
+}
+```
+
+Non-blocking hooks are the same as the after event hooks, response body will be ignored.
+
+
+### Naming convention
+
+For blocking hooks, we create them by different scenarios, so no specific naming convention.
+
+For non-blocking hooks, the main purposes are for data synchronization / app server notification, we will follow the pattern of `<entity>.<action>.<reason>`.
+
+
+### Proposed hooks
+
+#### Blocking hooks
+
+- pre_registration
+- admin_create_user
+
+#### Non-blocking hooks
+
+- user.created.user_signup
+- user.created.admin_api_create
+- identity.created.user_create
+- identity.created.admin_api_create
+- identity.deleted.user_delete
+- identity.deleted.admin_api_delete
+- identity.updated.user_update
+- session.created.user_signup
+- session.created.user_login
+- session.deleted.user_revoke
+- session.deleted.user_logout
+- session.deleted.admin_api_revoke
+- user.promoted.user_promote
+
+### Webhook Event hooks shape
+
+The current webhook shape looks good to me, i want to keep it with some minor update only.
+
+Quick preview of event:
+
+```json5
+{
+  "id": "0E1E9537-DF4F-4AF6-8B48-3DB4574D4F24",
+  "seq": 435,
+  "type": "after_user_create",
+  "payload": { /* ... */ },
+  "context": { /* ... */ }
+}
+```
+
+1. `type` will be the new event type
+1. Add `locale` to context
+1. Remove `reason` from `payload`
+
+
+Please check the below list for webhook shape
+- [Webhook shape](./webhook.md#webhook-event-shape)
+- [Payload of events](./webhook.md#blocking-webhook-event-list)
+
+
+### Discussion questions
+
+1. Seems we cannot solve developers may missed some of the events after we add new events problem, unless we support subscribing event type `user.created.*`. But we cannot do it in my blocking hooks definition, so eventually they will miss. I found the existing services don't have this problem, since they didn't put the reason in the event name, even in the request payload. Developer has no way to know the reason why the event is triggered. e.g.
```

Comment:
> Louis: Do we follow Stripe and send all events to a single endpoint? If yes, then the developer can actually match the event type with a prefix like user.created. provided that they do not care about the reason.

I think the major problems is blocking hooks, stripe doesn't have blocking events. If we allow user subscribe all blocking hooks in one endpoint, the default implementation of his webhook will be `is_allowed` = true? There are some problems that I can think of:

1. There are some performance tradeoff to have many unnecessary for **blocking hooks**, do we think it is acceptable?
2. We may need to have naming convention so developer can match prefix, but the naming need to be flexible enough. e.g. Ben mentioned example `after_signup_filled_email`.

If we want to go for this direction, we can think about how to solve point 2.

Reply:
Actually my comment is mainly about the non-blocking events.

For blocking events, I think the following properties are important to the developer.

- If the developer has defined a blocking hook for a particular event, then it is very likely that they want to know about the event and take actions. For example, if the developer has defined a blocking hook to inspect the email, it is very likely that they want to know about ALL scenarios. If our design could make them miss the event, it would be very surprising.
- If the developer has defined a blocking hook for a particular event, then the developer must explicitly tell us to allow the operation. Otherwise, the operation is allowed by default.